pex/pex_builder.py

```diff
@@ -784,24 +847,87 @@ def _build_zipapp(
-                # attempt to zip up our chroot. Bytecode compilation produces ephemeral temporary
-                # pyc files that we should avoid copying since they are useless and inherently
-                # racy.
-                exclude_file=is_pyc_temporary_file,
-                compress=compress,
-            )
+            # We will be reusing the same file handle to wrap with ZipFile, which needs to be able
+            # to read the file contents when instantiated, so it can check for existing zip
+            # file data. Hence, we open with "w+b".
+            with safe_io_open(filename, "w+b") as pexfile:
+                # (1) Write shebang line and any header script to a non-zip file handle.
+                pexfile.write(to_bytes("{}\n".format(self._shebang)))
+                if self._header:
+                    pexfile.write(to_bytes(self._header))
+
+                # (2) Zip up everything that won't be retrieved from a cache first.
+                # NB: This produces a slightly different output than if we had simply zipped up the
+                #     entire chroot at once; the sources and resources will now be listed before the
+                #     contents of the .bootstrap/ directory. This isn't necessary, but it makes the
+                #     output of `unzip -l` more useful for quickly scanning the contents of a PEX.
+                with TRACER.timed("zipping up uncached sources", V=9):
+                    # Reuse the file handle to zip into. This isn't necessary (we could close and
+                    # reopen it), but it avoids unnecessarily flushing to disk.
+                    self._chroot.zip(
+                        pexfile,
+                        mode="a",
+                        # These files are the only ones that might have a nonzero timestamp. All
+                        # entries copied from cache are assigned the zero timestamp, so their
+                        # checksums won't depend on the state of the pex cache.
+                        deterministic_timestamp=deterministic_timestamp,
+                        compress=compress,
+                        labels=self._DIRECT_SOURCE_LABELS,
+                    )
+
+                # (3) Engage our hacky solution to reuse the same caches as for --layout packed.
+                #
+                #     This reads from a single compressed file per cached resource instead of
+                #     traversing any directory trees. If compress=True, this will also significantly
+                #     reduce the total number of bytes to read and avoid a lot of single-threaded
+                #     computation, because the compressed entries are simply copied over as-is.
+                #
+                #     Note that these cached zips are created in the --layout packed format, without
+                #     the .bootstrap/ or .deps/ prefixes we need to form a proper zipapp. Our zip
```

Comment:
"proper zipapp" - it's a quibble perhaps, but the directory structure is a pure (internal) Pex thing and has nothing to do with zipapps, which only really care about zip files, the Python zipimporter and `__main__.py`. Maybe "proper Pex zipapp layout".